azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,325 @@
+            states=states,
+            transitions=transitions,
+            initial="disconnected",
+            send_event=True,  # Use event_data structures to pass transition arguments
+            finalize_event=_on_transition_complete,
+            queued=True,
+        )
+
+    def register(self, callback=None):
+        """
+        Register the device with the provisioning service.
+        """
+        logger.info("register called from polling machine")
+        self._register_callback = callback
+        self._trig_register()
+
+    def cancel(self, callback=None):
+        logger.info("cancel called from polling machine")
+        self._cancel_callback = callback
+        self._trig_cancel()
+
+    def _initialize_register(self, event_data):
+        logger.info("Preparing for registration process.")
+        # Choice: Sending action here as the polling/dps state machine must figure out
+        # in what order action must be sent. The request response is only responsible for
+        # requesting and responding. However the actions can be taken down to the
+        # the request response machine as well.
+        subscribe_action = SubscribeAction(
+            constant.SUBSCRIBE_TOPIC_PROVISIONING, qos=1, callback=self.on_subscribe_completed
+        )
+        # The request response takes a tuple of actions
+        # It will call state based provider connect which
+        # It will add subscribe to pending action queue
+        # it will execute connect and then execute subscribe
+        # TODO : Should I have a waiting callback or a connack on which something happens ?
+        self._request_response_provider.connect(
+            self.on_connect_completed, post_connect_actions=[subscribe_action]
+        )
+
+    def _send_register_request(self, event_data):
+        logger.info("Sending registration request from polling machine")
+        self._set_query_timer()
+
+        rid = str(uuid.uuid4())
+
+        publish_action = PublishAction(
+            constant.PUBLISH_TOPIC_REGISTRATION + rid, " ", callback=self.on_publish_completed
+        )
+        self._operations[rid] = publish_action
+        self._request_response_provider.send_request(publish_action)
+
+    def _query_operation_status(self, event_data):
+        logger.info("Querying operation status from polling machine")
+        self._set_query_timer()
+
+        rid = str(uuid.uuid4())
+        registration_result = event_data.args[0].args[0]
+
+        operation_id = registration_result.operation_id
+        publish_action = PublishAction(
+            constant.PUBLISH_TOPIC_QUERYING_I
+            + rid
+            + constant.PUBLISH_TOPIC_QUERYING_II
+            + operation_id,
+            " ",
+            callback=self.on_publish_completed,
+        )
+        self._operations[rid] = publish_action
+        self._request_response_provider.send_request(publish_action)
+
+    def _handle_response_received(self, registration_result):
```

Comment:
>_handle_response_received [](start = 8, length = 25)

we have to be careful about threading here.  I'm guessing that this function gets called in the context of a Paho thread, and we're calling a bunch of functions inside that thread, and that can screw up paho if we're not careful.

Reply:
+1

---
In reply to: [277404302](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277404302) [](ancestors = 277404302)